pandas/core/groupby/groupby.py

```diff
@@ -3290,6 +3304,102 @@ def max(
+        numeric_only : bool, default False
+            Include only float, int, boolean columns.
+        skipna : bool, default True
+            Exclude NA/null values. If an entire row/column is NA, the result
+            will be NA.
+        ignore_unobserved : bool, default False
+            When True and an unobserved group is encountered, do not raise. This used
+            for transform where unobserved groups do not play an impact on the result.
+
+        Returns
+        -------
+        Series or DataFrame
+            idxmax or idxmin for the groupby operation.
+        """
+        # seen is passed to the Cython code and mutated
+        seen = np.zeros(self.ngroups, dtype="bool")
+
+        def post_process(res):
+            has_na_value = (res._values == -1).any(axis=None)
+            has_unobserved = not seen.all()
+            raise_err: bool | np.bool_ = not ignore_unobserved and has_unobserved
+            if (
+                not self.observed
+                and not ignore_unobserved
+                and not raise_err
+                and (len(self.grouper.groupings) > 1 or res.empty)
+            ):
+                # When there are multiple groupings or the result is empty, the
+                # categories are not included in the upfront computation so we
+                # compare the final result length with the current length
+                result_len = np.prod(
+                    [len(ping.group_index) for ping in self.grouper.groupings]
+                )
+                raise_err = len(res) < result_len
+            if raise_err:
+                raise ValueError(
+                    f"Can't get {how} of an empty group due to unobserved categories. "
+                    "Specify observed=True in groupby instead."
+                )
+            elif not skipna and has_na_value:
+                warnings.warn(
+                    f"The behavior of {type(self).__name__}.{how} with all-NA "
+                    "values, or any-NA and skipna=False, is deprecated. In a future "
+                    "version this will raise ValueError",
+                    FutureWarning,
+                    stacklevel=find_stack_level(),
+                )
+
+            index = self.obj._get_axis(self.axis)
+            if res.size == 0:
+                result = res.astype(index.dtype)
+            else:
+                if isinstance(index, MultiIndex):
+                    index = index.to_flat_index()
+                values = res._values
+                na_value = na_value_for_dtype(index.dtype, compat=False)
+                if isinstance(res, Series):
+                    result = res._constructor(
+                        index.array.take(values, allow_fill=True, fill_value=na_value),
+                        index=res.index,
+                        name=res.name,
+                    )
+                else:
+                    buf = {}
+                    for k, column_values in enumerate(values.T):
```

Comment:
Is .T not expensive here at all? Normally this would be a pretty expensive code path but might be missing the larger assumption around the data at this point

Reply:
values is a NumPy array, this is cheap.

Reply:
Ah OK. For some reason I thought `.T` produced a copy but that is not true.